daemon/cmd/daemon_main.go

```diff
@@ -1743,6 +1743,13 @@ func startDaemon(d *Daemon, restoredEndpoints *endpointRestoreState, cleaner *da
 		<-params.CacheStatus
 	}
 	bootstrapStats.k8sInit.End(true)
+
+	// Wait for the ipcache to write at least once before restoring.
+	// Note that this must happen after waiting for k8s to sync, as the ipcache
+	// also blocks until k8s caches have synced.
+	log.Info("Waiting for ipcache to complete initial write")
+	d.ipcache.WaitForFirstLabelInjection()
```

Comment:
AFAIU, this synchronous wait not only blocks the restoration but all of the following tasks in the `startDaemon` function - and therefore might unnecessarily increase the startup time of the agent.

Since the restoration itself is also executed asynchronously in its own go-routine(s), wouldn't it be nicer to use an the asynchronous mechanism (and signal via channel) to wait for the first ipcache write too? This way we would only make the dependent "tasks" waiting for this.

Reply:
Sure, makes total sense.